// @cloudflare/ai/dist/tasks/sentence-similarity.js
import { Tensor, TensorType } from "../tensor";
export class AiSentenceSimilarity {
    constructor(inputs) {
        this.schema = {
            input: {
                type: "object",
                properties: {
                    source: {
                        type: "string",
                    },
                    sentences: {
                        type: "array",
                        items: {
                            type: "string",
                        },
                    },
                },
                required: ["source", "sentences"],
            },
            output: {
                type: "array",
                contentType: "application/json",
                items: {
                    type: "number",
                },
            },
        };
        this.inputs = inputs;
    }
    preProcessing() {
        this.preProcessedInputs = this.inputs;
    }
    generateTensors() {
        this.tensors = [
            new Tensor(TensorType.String, [this.preProcessedInputs.source], {
                shape: [1],
                name: "source_sentence",
            }),
            new Tensor(TensorType.String, this.preProcessedInputs.sentences, {
                shape: [this.preProcessedInputs.sentences.length],
                name: "sentences",
            }),
        ];
    }
    postProcessing(response) {
        this.postProcessedOutputs = response.scores.value;
    }
}
